File: system-test/java/src/main/java/com/optivem/shop/testkit/driver/adapter/external/clock/ClockStubDriver.java
package com.optivem.shop.testkit.driver.adapter.external.clock;

import com.optivem.shop.testkit.driver.port.external.clock.ClockDriver;

import com.optivem.shop.testkit.driver.adapter.external.clock.client.ClockStubClient;
import com.optivem.shop.testkit.driver.adapter.external.clock.client.dtos.ExtGetTimeResponse;
import com.optivem.shop.testkit.driver.port.external.clock.dtos.GetTimeResponse;
import com.optivem.shop.testkit.driver.port.external.clock.dtos.ReturnsTimeRequest;
import com.optivem.shop.testkit.driver.port.shared.dtos.ErrorResponse;
import com.optivem.shop.testkit.common.Closer;
import com.optivem.shop.testkit.common.Converter;
import com.optivem.shop.testkit.common.Result;

public class ClockStubDriver implements ClockDriver {
    private final ClockStubClient client;

    public ClockStubDriver(String baseUrl) {
        this.client = new ClockStubClient(baseUrl);
    }

    @Override
    public void close() {
        client.removeStubs();
        Closer.close(client);
    }

    @Override
    public Result<Void, ErrorResponse> goToClock() {
        return client.checkHealth()
                .mapError(ext -> ErrorResponse.builder().message(ext.getMessage()).build());
    }

    @Override
    public Result<GetTimeResponse, ErrorResponse> getTime() {
        return client.getTime()
                .map(ext -> GetTimeResponse.builder().time(ext.getTime()).build())
                .mapError(ext -> ErrorResponse.builder().message(ext.getMessage()).build());
    }

    @Override
    public Result<Void, ErrorResponse> returnsTime(ReturnsTimeRequest request) {
        var time = Converter.toInstant(request.getTime());

        var extResponse = ExtGetTimeResponse.builder()
                .time(time)
                .build();

        return client.configureGetTime(extResponse)
                .mapError(ext -> ErrorResponse.builder().message(ext.getMessage()).build());
    }
}

File: system-test/typescript/src/testkit/driver/adapter/external/clock/clock-stub-driver.ts
import { Result, success, failure } from '../../../../common/result.js';
import { ClockErrorResponse, GetTimeResponse, ReturnsTimeRequest } from '../../../../common/dtos.js';
import { ClockDriver } from '../../../port/external/clock/clock-driver.js';
import { JsonWireMockClient } from '../../shared/wiremock/wiremock-client.js';

export class ClockStubDriver implements ClockDriver {
  private wireMock: JsonWireMockClient;

  constructor(private baseUrl: string) {
    this.wireMock = new JsonWireMockClient(baseUrl);
  }

  async goToClock(): Promise<Result<void, ClockErrorResponse>> {
    const response = await fetch(`${this.baseUrl}/health`);
    if (response.ok) return success(undefined);
    return failure({ message: `Clock stub not available: ${response.status}` });
  }

  async getTime(): Promise<Result<GetTimeResponse, ClockErrorResponse>> {
    const response = await fetch(`${this.baseUrl}/api/time`);
    if (response.ok) {
      const data = (await response.json()) as GetTimeResponse;
      return success(data);
    }
    return failure({ message: `Failed to get time: ${response.status}` });
  }

  async returnsTime(request: ReturnsTimeRequest): Promise<Result<void, ClockErrorResponse>> {
    await this.wireMock.stubGet('/clock/api/time', { time: request.time });
    return success(undefined);
  }

  async close(): Promise<void> {
    await this.wireMock.removeStubs();
  }
}
